app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -42,7 +43,7 @@ echo "#"
 if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     CONSUMER_RPM=$(mktemp --suffix .rpm)
 
-    curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
+    curl --insecure --output $CONSUMER_RPM <%= url %>/pub/katello-ca-consumer-latest.noarch.rpm
```

Comment:
Please use the host parameter instead:
https://github.com/Katello/katello/blob/0af4bbeba0346544c015a58f61ceb16e84ffb810/app/helpers/katello/katello_urls_helper.rb#L9

Reply:
But `host` argument from `subscription_manager_configuration_url` can't be `string`

```
<%= subscription_manager_configuration_url('foreman2.example.com') %>

Backtrace for 'Action failed' error (NoMethodError): undefined method `content_source' for "something.example.com":String
```

Reply:
Then perhaps you need to fix Katello first. There are plans to change change the mechanism and I don't want to hardcode locations in this template.

This is a bit tricky to explain with text and without diagams, but my concern is with routing services to their correct hostname. The use case that IMHO we should be able to support is where a Smart Proxy is multi homed. That means it has an internal interface (hostname + IP) on a private LAN that Foreman can talk to. Then there's a public interface (hostname + IP) on the client network.

Let's call the internal one `smart-proxy.lan.example.com` and the public one `smart-proxy.client.example.com`.

To support this, the template module has a setting `template_url` that you can set to `https://smart-proxy.client.example.com:8443`.

If we get to the consumer RPM, things are different. This is served by Apache. It may be configured differently. That's why there's a `subscription_manager_configuration_url` helper to give the correct URL. Sadly, this assumes it's running on the same hostname as the Smart Proxy. This mirrored in the installer where there is no code to set up Apache on the client hostname. If you don't use the helper, there is no guarantee that it'll actually work. The templates Smart Proxy may not even have the Pulp content module enabled.

Another problem you may run into is that the chosen organization and/or location has no permissions to that content proxy.

To move forward, I am thinking about solutions. It'll be harder to template, but I really think that passing the URL is the wrong solution. You really need a Smart Proxy instance. From there you can derive the correct information.

For example, the host registration module (https://github.com/theforeman/smart-proxy/pull/764) needs to be enabled before you can actually use it. There should be Foreman application code to get the correct URL. If the module is not enabled, you probably want to fall back to the real Foreman URL. This is particularly important for upgrade scenarios. We want to get to a scenario where you can upgrade Foreman independently from Smart Proxies. If the Smart Proxy doesn't have the module yet because it's too old, the registration shouldn't break.

Similarly, when you have the Proxy, you can also check if content is present and the organization and location are allowed to use it.

I realize this is asking for a lot more, but I think you should not underestimate how complex it is what you're trying to achieve.

Reply:
@jlsherrill and @ehelms please keep me honest here.

Reply:
I'm going to try to look at it from a workflow perspective to help me understand some of the nuance. Prerequisite information for Katello scenario:

 * Smart proxy runs on port 9090
 * Reverse proxy hosting consumer RPM is on port 8443
 * Assume a single public hostname for both: smart-proxy.external1.example.com
 * Assume templates running on same smart proxy as Pulp and Reverse Proxy

The user on a client would run:

```
curl -X GET "https://smart-proxy.external1.example.com:9090/register?url=https://smart-proxy.external1.example.com:8443" --user admin:changeme
```

I think generally this could be shortened to:

```
curl -X GET "https://smart-proxy.external1.example.com:9090/register" --user admin:changeme
```

1) I assume that URL here is not required as users should be able to safely assume that in the simple case our software is smart enough to register them to the hostname they are curling against without having to repeat that.

2) If the user is indicating where they want to register to, that is the user is specifying the URL themselves, then it should be safe to trust and use that URL.

3) The above holds true for a load balanced setup as well as we need to trust the URL that is input from the client and not attempt to derive it ourselves.

4) As of today, I do not believe we try to care if a host has the permission to register through a given smart proxy but we do care whether the persona registering the client has access to the org and location being specified (or if its an activation key then that handles everything nicely for us).

Reply:
> To support this, the template module has a setting template_url that you can set to 

Yes, this is why we want to allow `:url` parameter in the template, so users could setup `template_url` in Smart Proxy `template` module.

> If we get to the consumer RPM, things are different.

So the solution would be to keep `<%= subscription_manager_configuration_url %>` and not replace it with `<%= url %>/pub/...`

>To move forward, I am thinking about solutions. It'll be harder to template, but I really think that passing the URL is the wrong solution. You really need a Smart Proxy instance. From there you can derive the correct information.

Yeah I should mention that in the description. The `:url` parameter here is mainly for URL passed by Smart proxy `templates` module, see https://github.com/theforeman/smart-proxy/pull/763

Reply:
> > If we get to the consumer RPM, things are different.
> 
> So the solution would be to keep `<%= subscription_manager_configuration_url %>` and not replace it with `<%= url %>/pub/...`

IMHO it'd be best if the Smart Proxy name would be sent as part of the request. The best way to do that is to allow client side certificates to authenticate. Then you're guaranteed that the request securely comes from a known Smart Proxy. `SmartProxyAuth` is the controller that's used right now. I don't know how well that works if you want auth to be optional.

On the Foreman side you can then find the Smart Proxy object. From that object you can get all the right information.

The good thing is that the templates module on the Foreman side already configures the connection to send client certificates, so no modification is needed there.

With https://github.com/theforeman/smart-proxy/pull/767 you could also consider not sending the template_url but rather let Foreman deduce it. Without the patch it's also possible, but that costs an API call.

Reply:
> IMHO it'd be best if the Smart Proxy name would be sent as part of the request

Why to implement another feature when `templates` module already does that:
https://github.com/theforeman/smart-proxy/blob/aa6501f6b7c26c1e82e53dc8e40b8bd748aac14a/modules/templates/proxy_request.rb#L29

Reply:
Because they `template_url` is unrelated to the `subscription_manager_configuration_url`. One runs on some host, typically port 8000 and the other is served by Apache, typically on port 80. They can even have different hostnames.

Reply:
Yeh, I wrote that in previous comment, the solution would be to keep `<%= subscription_manager_configuration_url %>` and not replace it with `<%= url %>/pub/...`

I'll update the PR.

Reply:
The thing is that the templates URL may differ from the Smart Proxy name. In https://github.com/theforeman/smart-proxy/pull/763#issuecomment-687159646 I've described how this can show up.

In the case of Katello, it's important to know that `subscription_manager_configuration_url` by default points to the Foreman host. In an isolated setup where clients can't talk to Foreman and other services directly, they register to the Smart Proxy with content. That's why it can take a Smart Proxy parameter.

This is how we determine the actual Smart Proxy name using the client certificates:

https://github.com/theforeman/foreman/blob/c9ec02be2837eeb78323efcad0e9c1300e7535b6/app/controllers/concerns/foreman/controller/smart_proxy_auth.rb#L59-L103

Do note that if you include that controller, it counts as user authentication which is what we don't want here since username and password should be checked. That means the controller needs to be refactored.

The good news is that the requests sent by Templates already have the SSL client certificates set up for authentication:

https://github.com/theforeman/smart-proxy/blob/aa6501f6b7c26c1e82e53dc8e40b8bd748aac14a/lib/proxy/request.rb#L71-L83

Reply:
My understanding of @ehelms comment is, we should just trust the incoming url parameter. Proxy should be smart enough to provide reasonable default or user can override that, but that still happens on the proxy side. If no url parameter came, this is not "through-proxy" registration and we just use foreman url. For that to happen, we need a patch in katello to allow customizing [prefix of subscription_manager_configuration_url](https://github.com/Katello/katello/blob/0af4bbeba0346544c015a58f61ceb16e84ffb810/app/helpers/katello/katello_urls_helper.rb#L10). The logic on proxy (unrelated to this PR, will be addressed in https://github.com/theforeman/smart-proxy/pull/763) will work like this. Proxy will look if some explicit url was passed as a parameter. If yes, it passes it to Foreman, otherwise it parses it out of `request.env['REQUEST_URI']`. Therefore for any "through-proxy" registration, the url parameter in request to Foreman will be set. If client calls `https://proxy1.example.com:123/register` the url passed to Foreman will be `https://proxy1.example.com:123/`. If client calls `https://proxy1.example.com:123/register?url=https://proxy2.example.net` then the url ise set to `https://proxy2.example.net`. The url parameter is used for both host creation API request (without Katello) and consumer rpm downloading (with Katello). Note that it's always one or another case, not both at the same time. But we could go even further and have different parameters for url and content_url. That would be handy if user wants to perform the registration through one proxy and get the content from other.

There's no way to pass host to the `subscription_manager_configuration_url` method, the global registration template is rendered before the host even exist in our database. The host object does not exist at this point. If user want to set content source for newly registered host, they can specify the hostgroup, but it will only be set after the host is created.

I don't think we should be detecting the proxy by the cert and use it's URL at all. The URL that Foreman knows is the one that Foreman can resolve when it talks to the proxy. Client that's being registered does not have to be able to use such URL. We'll need this later, when we deal with deploying SSL certs for proxy, but that's not something we should solve in this PR.

I will consult this with @ehelms offline, but @ekohl if you have other concerns or want to discuss this more, let's have a short call, just ping me please.

Reply:
Also there's one inaccuracy repeated in few PRs.

> Do note that if you include that controller, it counts as user authentication which is what we don't want here since username and password should be checked.

This is not entirely truth. Just inlcuding the module does not do anything. There needs to be explicit `add_smart_proxy_filters` class method call that overrides the authentication logic, see e.g. https://github.com/theforeman/foreman/blob/82b4a36d0b52b6a9db3adf392595a3d6914afdc5/app/controllers/api/v2/config_reports_controller.rb#L11

In order to identify the proxy identity, we will need to split the [auth_smart_proxy](https://github.com/theforeman/foreman/blob/c9ec02be2837eeb78323efcad0e9c1300e7535b6/app/controllers/concerns/foreman/controller/smart_proxy_auth.rb#L59-L103) method, but that's easy task. We can still include this module and use just the identification part.